text/0088-metric-instrument-optional-refinements.md

```diff
@@ -0,0 +1,205 @@
+interest.  For a Sum-Only instrument refinement, we have a semantic
+property that two events with numeric values `M` and `N` are
+semantically equivalent to a single event with value `M+N`.  For
+example, in a count of users arriving by bus to an event, we are not
+concerned with the number of buses that arrived.
+
+A sum-only instrument is one where the number of events is not
+counted, by default, only the `Sum`.
+
+#### Language-level refinements
+
+OpenTelemetry implementations may wish to add instrument refinements
+to accommodate built-in types.  Languages with distinct integer and
+floating point should offer instrument refinements for each, leading
+to type names like `Int64Measure` and `Float64Measure`.
+
+A language with support for unsigned integer types may wish to create
+dedicated instruments to report these values, leading to type names
+like `UnsignedInt64Observer` and `UnsignedFloat64Observer`.  These
+would naturally apply a non-negative refinment.
+
+Other uses for built-in type refinements involve the type for duration
+measurements.  For example, where there is built-in type for the
+difference between two clock measurements, OpenTelemetry APIs should
+offer a refinement to automatically apply the correct unit of time to
+the measurement.
+
+### Counter refinement
+
+Counter is a non-negative, sum-only refinement of the Measure
+instrument.
+
+### Future refinements
+
+This leaves the potential to include other refinements by combining
+the above elements.  A `Counter`-like instrument that permits
+non-negative updates could be called an `UpDownCounter`, for example.
+An `Observer`-like instrument with non-descending values could be
+called a `MonotonicObserver` instrument. An `Observer`-like instrument
+that reports non-negative updates to a sum could be called a
+`DeltaObserver` instrument.
+
+## Internal details
+
+This is a change of understanding.  It does not request any new
+instruments be created or APIs be changed, but it does specify how we
+should think about adding new instruments.
+
+No API changes are called for in this proposal.
+
+### Note for Cumulative Exporters
+
+The Prometheus system collects cumulative data for its counter
+instruments, meaning it exports the lifetime sum of a counter at each
+collection interval, not the change in that sum over the collection
+interval. 
+
+It is important, therefore, to know when the output of an instrument
+should be reflected as a cumulative value in the exporter.  The
+`Counter` instrument automatically has this property, but why?
+
+We can infer that a value is cumulative in the following circumstances:
+
+- Non-negative and Sum-only: This is called `Counter`, if synchronous, and can be `DeltaObserver`, potentially, if asynchronous
+- Monotonic: This can be `MonotonicObserver`, potentially.
+
+We see that these refinements satisfy their intended purpose, which is
+to convey additional semantics that exporters use to convert data into
+their system.
```

Comment:
Interesting section here. If I'm running a prometheus backend, I would want my counter aggregations to be cumulative. If I'm running one that expects deltas to be reported for each interval, I'd want my counter aggregations to be non-cumulative (deltas). 

Do we foresee OTel providing a built-in way for an exporter to convert between the two? If not, how do we choose the default counter aggregation in a way that won't end up making someone's life difficult by having to build their own state into their exporter code, or, maybe worse, make the operator have to provide custom aggregation for all their counters?

Reply:
One possible solution: have count aggregators provide both a delta and a cumulative value, so exporters can choose appropriately the value that works for their backend.

Reply:
I think this should be a property of the Views (options to configure aggregations). I do agree that the default should be more like other cncf projects offer

Reply:
I don't see how the Views API will help an exporter author deal with having to map between cumulative counters & delta counters or vice-versa. Can you describe how this would work?

Reply:
I think I understand your concern, probably we should always build deltas from the Measure where we have full control and then the export like Prometheus can merge deltas. For observer where we don't have control of the values that are recorded they may be deltas or cumulative, probably we need to know that in the exporter.

Reply:
My goal was to let the "Sum-only" refinement imply that a sum was being monitored. For exporters that export cumulative sums, the Integrator should know to use a stateful aggregator. For exporters that export deltas, the Integrator should know to use a stateless aggregator.

The existing Go Integrator code has a boolean to indicate whether it is stateful or not, which appears to address this problem. For Prometheus, however, it doesn't do enough. The reason is that Summary-valued aggregators are supposed to be reported as deltas, whereas Histogram-valued aggregators are suppose to be reported as cumulative values. So we need a finer-grained way to select whether an aggregator is stateful, and it appears to depend on the exporter.

Therefore, I agree with @bogdandrutu that this is a problem best left to the Views API. Still, it's possible when there is only one Exporter to infer the correct default behavior for these Sum-only instruments, but this won't work if there are multiple exporters.

Reply:
@jmacd even for Histograms we can do deltas, they are easy to merge :)

Reply:
Should exporters have to declare what kind of sum-like aggregations they want (cumulative vs. deltas)? I'm not sure how else that information would get communicated. I'm definitely not understanding how the Views API would help here.

Reply:
Also, I think this discussion, although important, is not really relevant to this particular OTEP. :)